Adding BEM naming convention to the stylelint rules

rafaelgallani opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

## Description
Fixes https://github.com/WordPress/gutenberg/issues/28616. 
It's basically an increment to a `stylelint` rule validation regex, extending it to also allow BEM naming convention (`block__element--modifier`).

## How has this been tested?
`npx stylelint **/*.scss` and making sure tests wouldn't break.
 
## Types of changes
Bugfix - I guess. Since the validation was disabled...
I'm not sure how to categorize it. 

<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/native-mobile.md -->

rafaelgallani committed:
Adding BEM naming convention to the stylelint rules

rafaelgalani commented:
@gziolo the checks have failed since the rule is now applied in the linter. What should I do in this case, disable the rule and keep only the regex change, or change the CSS rules as well? 😅

ocean90 commented on packages/stylelint-config/index.js:
```diff
@@ -98,7 +98,7 @@ module.exports = {
 		'selector-attribute-quotes': 'always',
 		'selector-class-pattern': [
-			'^([a-z][a-z0-9]*)(-[a-z0-9]+)*$',
+			'^([a-z][a-z0-9]*)(-[a-z0-9]+)*((__([a-z][a-z0-9]*)(-[a-z0-9]+)*)?(--([a-z][a-z0-9]*)(-[a-z0-9]+)*)?)$',
```
I suggest to also set the [`resolveNestedSelectors ` option](https://stylelint.io/user-guide/rules/selector-class-pattern#resolvenestedselectors-true--false-default-false) and the message needs an update too. See [this example](https://github.com/wearerequired/coding-standards/blob/68ce2de9a3a723146afb3bb681d363ae869a7388/packages/stylelint-config/index.js#L31-L38) for inspiration.

gziolo commented:
Nice one, the number of violations reported is so long 😅

> the checks have failed since the rule is now applied in the linter. What should I do in this case, disable the rule and keep only the regex change, or change the CSS rules as well? 😅

We definitely should ignore those violations for React Native files (.native.scss, .ios.scss, .android.scss)

rafaelgalani commented:
> We definitely should ignore those violations for React Native files (`.native.scss`, `.ios.scss`, `.android.scss`). However, I couldn't find a way for overrides on a subset of files in stylelint.

@gziolo Gotcha. There's probably an option... I tried it on #28988. 
Basically: `stylelint` allows you to disable specific rules for a file/line with a comment, just like ESLint does. 
What I tried was creating a plugin that would enforce the mobile SCSS files to have this comment line as their first one, so the rule would be entirely disabled on them. Is it too hacky? 😅 
Also, after applying the rule, there were only 222 occurrences of the error, which is way less than the previous 738!!! 🎉🎉

gziolo commented:
@gziolo Gotcha. There's probably an option... I tried it on #28988.
> Basically: `stylelint` allows you to disable specific rules for a file/line with a comment, just like ESLint does.
> What I tried was creating a plugin that would enforce the mobile SCSS files to have this comment line as their first one, so the rule would be entirely disabled on them. Is it too hacky? 😅

I commented on the PR and asked the mobile team to share their thoughts on the approach proposed. Great idea to automate it, it would be too much hassle otherwise 😄 

> Also, after applying the rule, there were only 222 occurrences of the error, which is way less than the previous 738!!! 🎉🎉

It sounds promising. We should also disable this rule in https://github.com/WordPress/gutenberg/blob/master/packages/components/src/date-time/datepicker.scss since we forked https://github.com/airbnb/react-dates so those class names are completely different. With that, the number of violations should be way smaller.

dcalhoun committed:
Scope selector-class-pattern stylelint rule to web stylesheets (#29881)

* Scope selector-class-pattern stylelint rule to web stylesheets

Native stylesheets currently do not adhere to the same BEM selector
patterns as web stylesheets. This introduces a custom stylelint plugin
to exclude native stylesheets from the selector-class-pattern rule.

* Fixing unit tests to use correct base directory

* Append native suffix to avoid web-specific lint rules

All stylesheet files that do not have the native suffix are linted with
web-specific selector rules. Adding the suffix disables those irrelevant
rules for the native stylesheet.

Co-authored-by: Rafael Galani <galani.rafael@gmail.com>

gziolo commented:
I merged work done by @dcalhoun in #29881 to this branch. Let's see what's still reported by CI and how we can move forward.

Mamaduka commented:
Hi, @rafaelgalani

The original issue came up during Weekly Editor Bug Scrub.

I just wanted to check if you've time to continue working on this PR or how I can help to move this forward.

Thanks

gziolo commented:
It would be great to move forward with this PR, just saying 😄
